{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Normal protein sequence:\n",
      "<!DOCTYPE html PUBLIC \"-//W3C//DTD XHTML 1.0 Strict//EN\" \"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd\">\n",
      "<html xmlns=\"http://www.w3.org/1999/xhtml\" xml:lang=\"en\" lang=\"en\">\n",
      "  <head>\n",
      "    <meta http-equiv=\"Content-Type\" content=\"text/html; charset=utf-16\" />\n",
      "    <meta name=\"keywords\" content=\"SIFT, non-synonymous polymorphism, nonsynonymous polymorphisms, Single Nucleotide Polymorphims analysis tool, mutation analysis tool, protein function prediction tool, amino acid substitution effect prediciton tool, human SNPs, human polymorphisms, consequence of amino acid changes, missense mutations, nsSNP, polymorphism, polymorphisms, human genetic variations, human mutagenesis, SNPs analysis tool, genome variant, genomic variation, mutation, coding nonsynonymous variants, effects of amino acid substitutions, affect protein function, deleterious AAS, Deleterious Amino Acid Substitutions, CDS variants, Damaging Amino Acid Substitutions, single nucleotide variants, genome-scale, SIFT predictions, SIFT scores, missense status, in-silico prediction, protein sequence conservation, protein sequence homology, protein homologues, multiple alignment, ng, henikoff, sorting intolerant from tolerant, SNP, bii, gis, bioinformatics, computational biology, SIFT genome tools, single protein tools, protein batch tools, automated prediction tool, algorithm, genotype change, fhcrc, jcvi, Fred Hutchinson Cancer Researcher Center, J. Craig Venter Institute, Genome Institute of Singapore, Bioinformatics Institute, protein FASTA sequence\"/>\n",
      "\n",
      "    <link rel=\"stylesheet\" type=\"text/css\" href=\"../css/sift.css\" />\n",
      "    <link rel=\"shortcut icon\" href=\"/images/favicon.ico\">\n",
      "\n",
      "<!-- stylesheets need to be named >\n",
      "    <!--[if IE]> -->\n",
      "      <style type=\"text/css\">\n",
      "        #content { overflow: hidden; width: 933px; } \n",
      "        #main_content_3c #content, #main_content_2c #content { width: 953px; } \n",
      "        #main_content_2c #content #middle_content { width: 710px; }\n",
      "        #main_content_3c #middle_content { width: 515px; }\n",
      "      </style>\n",
      "    <!--[endif]-->\n",
      "    <!--[if lt IE 7.]-->\n",
      "      <style type=\"text/css\">\n",
      "        #main_logo { height: 79px; }\n",
      "      </style>\n",
      "    <!--[endif]-->\n",
      "    <title>SIFT Single Protein Sequence</title>\n",
      "    <script src=\"https://code.jquery.com/jquery-3.5.1.js\"></script>\n",
      "    <script src=\"../javascripts/webform_checks.js\"></script>\n",
      "\n",
      "    <link rel=\"stylesheet\" type=\"text/css\" href=\"../css/sift.css\" />\n",
      "    <link rel=\"stylesheet\" type=\"text/css\" href=\"../css/refresh.css\"/>\n",
      "    <link rel=\"shortcut icon\" href=\"/images/favicon.ico\">\n",
      "    <script type=\"text/javascript\" charset=\"UTF-8\" src=\"/javascripts/refresh.js\"> </script>\n",
      "\n",
      "<script type=\"text/javascript\">\n",
      "function ChgText()\n",
      "{\n",
      "   var MyElement = document.getElementById(\"queryseq\");\n",
      "   MyElement.value = \n",
      "\">p53 isoform a Homo sapiens\\n\" \n",
      "+ \"MEEPQSDPSVEPPLSQETFSDLWKLLPENNVLSPLPSQAMDDLMLSPDDIEQWFTEDPGPDEAPRMPEAA\"  \n",
      "+ \"PPVAPAPAAPTPAAPAPAPSWPLSSSVPSQKTYQGSYGFRLGFLHSGTAKSVTCTYSPALNKMFCQLAKT\" \n",
      "+ \"CPVQLWVDSTPPPGTRVRAMAIYKQSQHMTEVVRRCPHHERCSDSDGLAPPQHLIRVEGNLRVEYLDDRN\" \n",
      "+ \"TFRHSVVVPYEPPEVGSDCTTIHYNYMCNSSCMGGMNRRPILTIITLEDSSGNLLGRNSFEVRVCACPGR\" \n",
      "+ \"DRRTEEENLRKKGEPHHELPPGSTKRALPNNTSSSPQPKKKPLDGEYFTLQIRGRERFEMFRELNEALEL\" \n",
      "+ \"KDAQAGKEPGGSRAHSSHLKSKKGQSTSRHKKLMFKTEGPDSD\"; \n",
      "\n",
      "  var MyElement2 = document.getElementById (\"substitutions\");\n",
      "  MyElement2.value = \"Q5H\\nD7H # somatic mutation\\nS46A # abolishes phosphorylation\\nP359D #no binding to USP7\" ; \n",
      "}\n",
      "\n",
      "//--></script>\n",
      "\n",
      "    <style type=\"text/css\">\n",
      "      #middle_content_template { \n",
      "        margin: 0;\n",
      "        margin-left: 50px;\n",
      "      } /* end #middle_content_template */\n",
      "      \n",
      "      #middle_content_template, #left_content ul { text-align: left; }\n",
      "    </style>\n",
      "\n",
      "    \n",
      "    \n",
      "  </head>\n",
      "  <body> \n",
      "    <div class=\"modal-overlay\"></div>\n",
      "    <div style=\"position:fixed; z-index:102; top:50%; left:50%;\">\n",
      "      <div class=\"loading\" id=\"modal-box\">\n",
      "          <img class=\"loading\" id=\"ajax-img\" src=\"../images/ajax-progress-bar-blue.gif\"/>\n",
      "          <div class=\"loading\" id=\"ajax-text\" >\n",
      "                <p>Processing your SIFT submission ...</p>\n",
      "          </div>\n",
      "          <img class=\"loading\" id=\"ajax-close\" src=\"../images/ajax-cancel.png\" />\n",
      "      </div>\n",
      "    </div>\n",
      "\n",
      "    <div id=\"logo_search\" class=\"clearfix\">\n",
      "      <div id=\"logo_container\">\n",
      "        <div id=\"search\">\n",
      "        </div>\n",
      "\n",
      "  <script src=\"/javascripts/heading.js\" type=\"text/javascript\"> </script>\n",
      "\n",
      "    <div id=\"menu_border\">\n",
      "      <div id=\"menu1\">\n",
      "        <div id=\"menu_container\">\n",
      "          <ul>\n",
      "          </ul>\n",
      "        </div>\n",
      "      </div>\n",
      "    </div>\n",
      "    <div id=\"content\">\n",
      "      <div id=\"container_1\">\n",
      "        <div id=\"container_2\">\n",
      "\n",
      "                </ul>\n",
      "\n",
      "              </div>\n",
      "\t      <FORM ENCTYPE=\"MULTIPART/FORM-DATA\" METHOD=\"POST\" ACTION=\"/sift-bin/SIFT_seq_submit2.pl\" id=\"Form\">\n",
      "              <div id=\"middle_content_template\">\n",
      "\t        <div id=\"tool_instructions\">\n",
      "\t\t  <div id=\"instruct_title\">SIFT Sequence</div>\n",
      "\n",
      "                  <p><b>SIFT Sequence provides SIFT predictions for a given protein FASTA sequence.</b> This will take 10-15 min because we must search your protein sequence against a database to pick the related sequences. You can also <A HREF=\"SIFT_related_seqs_submit.html\">submit your protein sequence and related sequences</A> or <A HREF=\"SIFT_aligned_seqs_submit.html\">aligned sequences </A> if you already have them.<BR>\n",
      " \n",
      "\t\t\t<BR>\n",
      "\t\t\t<i>Results are deleted after 24 hours, so please save them! </i>\n",
      "\t\t\t<BR>\n",
      "\t\t\t[<A HREF=\"SIFT_help.html#CHANGING_PARAMETERS\">Preventing connection failures</A>]\n",
      "\n",
      "\n",
      "\t\t  </p>\n",
      "\n",
      "\t\t</div>\n",
      "\n",
      "\t\t\t\n",
      "\t\t\t<HR>\n",
      "\t\t\t<font color=blue>Protein sequence</font><BR>\n",
      "\t\t\t<BR>\n",
      "\t\t\tName of file containing protein query sequence\n",
      "\t\t\t<A HREF=\"http://www.ncbi.nlm.nih.gov/BLAST/fasta.shtml\">(fasta format)</A>.<BR>\n",
      "\t\t\t<INPUT SIZE=50 NAME=\"query_file\" TYPE=\"file\"><BR><BR>\n",
      "\t\t\t<center><font size=\"6\" color=\"blue\">-or-</font></center><BR>\n",
      "\t\t\tPaste in your protein query sequence \n",
      "\t\t\t(<A HREF=\"javascript:ChgText()\" >Upload example</A>)\n",
      "\t\t\t<A HREF=\"http://www.ncbi.nlm.nih.gov/BLAST/fasta.shtml\">(fasta format)</A>.<BR>\n",
      "\t\t\t<TEXTAREA NAME = \"queryseq\" id=\"queryseq\" ROWS=10 COLS=70></TEXTAREA>\n",
      "\t\t\t<BR>\n",
      "\t\t\t<BR>\n",
      "\t\t\t<HR>\n",
      "\t\t\t<BR>\n",
      "\t\t\tEnter the substitutions of interest <A HREF=\"SIFT_help.html#substitution\">[format]</A>:<BR>\n",
      "\t\t\t<TEXTAREA NAME = \"substitutions\" id=\"substitutions\" ROWS=5 COLS=50></TEXTAREA>\n",
      "\t\t\t<BR><font size=\"6\" color=\"blue\">-or-</font>\n",
      "\n",
      "\n",
      "\t\t\t<BR>\n",
      "\t\t\tUpload a file containing substitutions of interest <A HREF=\"SIFT_help.html#substitution\">[format]</A>:<br>\n",
      "                        <INPUT SIZE=50 NAME=\"substitution_file\" TYPE=\"file\"><BR><BR>\n",
      "\n",
      "\t\t\t<HR>\n",
      "\t\t\t<BR>\n",
      "\t\t\t<font color=blue>Parameters</font><BR>\n",
      "\t\t\t<BR>\n",
      "\t \t\tSelect database to search:\n",
      "  \t\t\t<SELECT NAME=\"database\" SIZE=\"1\">\n",
      "  \t\t\t<OPTION VALUE=\"sp\" SELECTED>UniProt-SwissProt 2010_09 \n",
      "  \t\t\t<OPTION VALUE=\"sp_tr\">UniProt-SwissProt + TrEMBL 2010_09 \n",
      "  \t\t\t<OPTION VALUE=\"ur_ninety\">UniRef90 2011_08\n",
      "<!-- \t\t\t<OPTION VALUE=\"nr\">NCBI nonredundant 2010 Sept --> \n",
      "  \t\t\t</SELECT>\n",
      "\t\t\t<BR><A HREF=\"SIFT_help.html#CHOOSING_DATABASE\">Warnings for you predictions? Change this option</A><BR>\n",
      "\t\t\t<BR>\n",
      "\t\t\t<BR>\n",
      "\t\t\tMedian conservation of sequences <INPUT TYPE=\"TEXT\" SIZE=5 NAME=\"info\" VALUE=3.00> <BR>\n",
      "\t\t\t<i>(between 0 and 4.32, we recommend between 2.75 and 3.25)</i><BR>\n",
      "\t\t\t<BR>\n",
      "\t\t\tRemove sequences more than\n",
      "\t\t\t<input type=text name=seq_identity_filter size=3 maxlength = 3 value=\"90\">\n",
      "\t\t\tpercent identical to query<BR>\n",
      "\t\t\t<i>This cleans out any polymorphic alleles (sequences in the database that are nearly identical to your protein but already containing the substitutions of interest). The contaminating sequences may cause the substitution to be predicted as tolerated.</i>\n",
      " <p>\n",
      "                        Enter your email address if you want the results through email : (Optional)<BR>\n",
      "\n",
      "                        <INPUT SIZE=50 NAME=\"address\">\n",
      "                        </INPUT SIZE>\n",
      "                        <BR>\n",
      "<HR>\n",
      "\t\t\t<BR>\n",
      "\t\t\t<!--<INPUT TYPE=submit>-->\n",
      "                <input type=\"button\" id=\"submit_button\" value=\"Submit\"></input>\n",
      "\n",
      "                <INPUT TYPE=reset>\n",
      "\n",
      "\n",
      "\t\t  </p>\n",
      "\t\t  </FORM>\n",
      "              </div>\n",
      "            </div>\n",
      "          </div>\n",
      "\n",
      "        </div>\n",
      "      </div>\n",
      "    </div>\n",
      "    <div id=\"home_border\">\n",
      "      <div id=\"home_bar\">\n",
      "        <div id=\"home_container\">\n",
      "        </div>\n",
      "      </div>\n",
      "\n",
      "      <div id=\"footer_spacer\"></div>\n",
      "    </div>\n",
      "\n",
      "<P>\n",
      "  <script src=\"/javascripts/footer.js\" type=\"text/javascript\"> </script>\n",
      "\n",
      "  </body>\n",
      "</html>\n",
      "\n",
      "<script>\n",
      "$( document ).ready(function() {\n",
      "       $(\"div.modal-overlay\").hide();\n",
      "      $(\".loading\").hide();\n",
      "});\n",
      "\n",
      "$('#submit_button').on('click', function(e) {\n",
      "    \n",
      "  \n",
      "     var tag_inputs = document.getElementsByTagName('input');\n",
      "     var textarea = document.getElementsByTagName ('textarea');\n",
      "     //inputs = inputs.concat (textarea);\n",
      "     var inputs = [];\n",
      "     for (var index = 0; index < tag_inputs.length; index++) {\n",
      "        inputs.push (tag_inputs[index]);\n",
      "     }\n",
      "     for (var index = 0; index < textarea.length; index++) {\n",
      "        inputs.push (textarea[index]);\n",
      "     }\n",
      "\n",
      "\n",
      "     for(var x = 0; x < inputs.length; x++) {\n",
      "           \n",
      "       // alert (\"checking \" + inputs[x].name);\n",
      "             if (inputs[x].name == \"query_file\") {\n",
      "                 if (!checkFileOK (inputs[x].value, ['.fasta', '.fa'], false)) {\n",
      "                    alert (\"File must end in .fasta or .fa\");\n",
      "                     return false;\n",
      "                 }\n",
      "              //   if (inputs[x].files.length > 0 && inputs[x].files[0].size > 5000000) {\n",
      "                  if (!checkFileSize (inputs[x].files)) {\n",
      "                      alert (\"Files should be < 5 MB.\");\n",
      "                      return false;\n",
      "                 }\n",
      "\n",
      "             } else if (inputs[x].name == \"queryseq\") {\n",
      "                // alert (\"check QuerySeq\" + inputs[x].value);\n",
      "                 if (!validateFasta (inputs[x].value)) {\n",
      "                     alert (\"Please put your sequence in fasta format\");\n",
      "                     return false;\n",
      "                 }\n",
      "             } else if (inputs[x].name == \"substitutions\") {\n",
      " //                alert (\"check substitutions\");\n",
      "                 if (!validateSubstitutions (inputs[x].value)) {\n",
      "                     alert (\"The substitutions are not in correct format\");\n",
      "                     return false;\n",
      "                 }\n",
      "             } else if (inputs[x].name == \"substitution_file\") {\n",
      "                    if (!checkFileOK (inputs[x].value, ['.subst', '.txt'], false)) {\n",
      "                        alert (\"File must end in .subst or .txt.  File size cannot exceed 5 MB.\");\n",
      "                     return false;\n",
      "                    }\n",
      "                    if ( !checkFileSize (inputs[x].files) ) {\n",
      "                        alert (\"Files should be < 5 MB.\");\n",
      "                        return false;\n",
      "                    }\n",
      "             } else if (inputs[x].name == \"info\") {\n",
      "                 if (!validateMedianSeqInfo (inputs[x].value, true)) {\n",
      "                     alert (\"Please set the median conservation of sequences between 0 and 4.32  . How about 3.00?\");\n",
      "                     return false;\n",
      "                 }\n",
      "                 //alert (\"check substitutions\"); // between 0 and 4.32 (this is a dable)\n",
      "             } else if (inputs[x].name == \"seq_identity_filter\") {\n",
      "                // alert (\"check substitutions\"); // string between 0 and 1[[\n",
      "                if (!validateSeqIdentity (inputs[x].value, true)) {\n",
      "                    alert (\"Sequence identity must be between 0 and 100. How about 90?\")\n",
      "                    return false;\n",
      "                }\n",
      "             } else if (inputs[x].name == \"address\") {\n",
      "                if (!validateEmail (inputs[x].value)) {\n",
      "                     alert (\"Please correct your email address\");\n",
      "                     return false;\n",
      "                 }\n",
      "             }\n",
      "          \n",
      "     } // end for loop\n",
      "     //alert (document.getElementById (inputs[x].name).validationMessage);\n",
      "\n",
      "//     $(\"div.modal-overlay\").show();\n",
      " //    $(\".loading\").show();\n",
      "\n",
      "\n",
      "     $(\"#Form\").submit();\n",
      "\n",
      "    \n",
      "\n",
      "       e.preventDefault();\n",
      " });\n",
      "</script>\n",
      "\n",
      "\n",
      "\n",
      "SNP-containing protein sequence:\n",
      "<!DOCTYPE html PUBLIC \"-//W3C//DTD XHTML 1.0 Strict//EN\" \"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd\">\n",
      "<html xmlns=\"http://www.w3.org/1999/xhtml\" xml:lang=\"en\" lang=\"en\">\n",
      "  <head>\n",
      "    <meta http-equiv=\"Content-Type\" content=\"text/html; charset=utf-16\" />\n",
      "    <meta name=\"keywords\" content=\"SIFT, non-synonymous polymorphism, nonsynonymous polymorphisms, Single Nucleotide Polymorphims analysis tool, mutation analysis tool, protein function prediction tool, amino acid substitution effect prediciton tool, human SNPs, human polymorphisms, consequence of amino acid changes, missense mutations, nsSNP, polymorphism, polymorphisms, human genetic variations, human mutagenesis, SNPs analysis tool, genome variant, genomic variation, mutation, coding nonsynonymous variants, effects of amino acid substitutions, affect protein function, deleterious AAS, Deleterious Amino Acid Substitutions, CDS variants, Damaging Amino Acid Substitutions, single nucleotide variants, genome-scale, SIFT predictions, SIFT scores, missense status, in-silico prediction, protein sequence conservation, protein sequence homology, protein homologues, multiple alignment, ng, henikoff, sorting intolerant from tolerant, SNP, bii, gis, bioinformatics, computational biology, SIFT genome tools, single protein tools, protein batch tools, automated prediction tool, algorithm, genotype change, fhcrc, jcvi, Fred Hutchinson Cancer Researcher Center, J. Craig Venter Institute, Genome Institute of Singapore, Bioinformatics Institute, protein FASTA sequence\"/>\n",
      "\n",
      "    <link rel=\"stylesheet\" type=\"text/css\" href=\"../css/sift.css\" />\n",
      "    <link rel=\"shortcut icon\" href=\"/images/favicon.ico\">\n",
      "\n",
      "<!-- stylesheets need to be named >\n",
      "    <!--[if IE]> -->\n",
      "      <style type=\"text/css\">\n",
      "        #content { overflow: hidden; width: 933px; } \n",
      "        #main_content_3c #content, #main_content_2c #content { width: 953px; } \n",
      "        #main_content_2c #content #middle_content { width: 710px; }\n",
      "        #main_content_3c #middle_content { width: 515px; }\n",
      "      </style>\n",
      "    <!--[endif]-->\n",
      "    <!--[if lt IE 7.]-->\n",
      "      <style type=\"text/css\">\n",
      "        #main_logo { height: 79px; }\n",
      "      </style>\n",
      "    <!--[endif]-->\n",
      "    <title>SIFT Single Protein Sequence</title>\n",
      "    <script src=\"https://code.jquery.com/jquery-3.5.1.js\"></script>\n",
      "    <script src=\"../javascripts/webform_checks.js\"></script>\n",
      "\n",
      "    <link rel=\"stylesheet\" type=\"text/css\" href=\"../css/sift.css\" />\n",
      "    <link rel=\"stylesheet\" type=\"text/css\" href=\"../css/refresh.css\"/>\n",
      "    <link rel=\"shortcut icon\" href=\"/images/favicon.ico\">\n",
      "    <script type=\"text/javascript\" charset=\"UTF-8\" src=\"/javascripts/refresh.js\"> </script>\n",
      "\n",
      "<script type=\"text/javascript\">\n",
      "function ChgText()\n",
      "{\n",
      "   var MyElement = document.getElementById(\"queryseq\");\n",
      "   MyElement.value = \n",
      "\">p53 isoform a Homo sapiens\\n\" \n",
      "+ \"MEEPQSDPSVEPPLSQETFSDLWKLLPENNVLSPLPSQAMDDLMLSPDDIEQWFTEDPGPDEAPRMPEAA\"  \n",
      "+ \"PPVAPAPAAPTPAAPAPAPSWPLSSSVPSQKTYQGSYGFRLGFLHSGTAKSVTCTYSPALNKMFCQLAKT\" \n",
      "+ \"CPVQLWVDSTPPPGTRVRAMAIYKQSQHMTEVVRRCPHHERCSDSDGLAPPQHLIRVEGNLRVEYLDDRN\" \n",
      "+ \"TFRHSVVVPYEPPEVGSDCTTIHYNYMCNSSCMGGMNRRPILTIITLEDSSGNLLGRNSFEVRVCACPGR\" \n",
      "+ \"DRRTEEENLRKKGEPHHELPPGSTKRALPNNTSSSPQPKKKPLDGEYFTLQIRGRERFEMFRELNEALEL\" \n",
      "+ \"KDAQAGKEPGGSRAHSSHLKSKKGQSTSRHKKLMFKTEGPDSD\"; \n",
      "\n",
      "  var MyElement2 = document.getElementById (\"substitutions\");\n",
      "  MyElement2.value = \"Q5H\\nD7H # somatic mutation\\nS46A # abolishes phosphorylation\\nP359D #no binding to USP7\" ; \n",
      "}\n",
      "\n",
      "//--></script>\n",
      "\n",
      "    <style type=\"text/css\">\n",
      "      #middle_content_template { \n",
      "        margin: 0;\n",
      "        margin-left: 50px;\n",
      "      } /* end #middle_content_template */\n",
      "      \n",
      "      #middle_content_template, #left_content ul { text-align: left; }\n",
      "    </style>\n",
      "\n",
      "    \n",
      "    \n",
      "  </head>\n",
      "  <body> \n",
      "    <div class=\"modal-overlay\"></div>\n",
      "    <div style=\"position:fixed; z-index:102; top:50%; left:50%;\">\n",
      "      <div class=\"loading\" id=\"modal-box\">\n",
      "          <img class=\"loading\" id=\"ajax-img\" src=\"../images/ajax-progress-bar-blue.gif\"/>\n",
      "          <div class=\"loading\" id=\"ajax-text\" >\n",
      "                <p>Processing your SIFT submission ...</p>\n",
      "          </div>\n",
      "          <img class=\"loading\" id=\"ajax-close\" src=\"../images/ajax-cancel.png\" />\n",
      "      </div>\n",
      "    </div>\n",
      "\n",
      "    <div id=\"logo_search\" class=\"clearfix\">\n",
      "      <div id=\"logo_container\">\n",
      "        <div id=\"search\">\n",
      "        </div>\n",
      "\n",
      "  <script src=\"/javascripts/heading.js\" type=\"text/javascript\"> </script>\n",
      "\n",
      "    <div id=\"menu_border\">\n",
      "      <div id=\"menu1\">\n",
      "        <div id=\"menu_container\">\n",
      "          <ul>\n",
      "          </ul>\n",
      "        </div>\n",
      "      </div>\n",
      "    </div>\n",
      "    <div id=\"content\">\n",
      "      <div id=\"container_1\">\n",
      "        <div id=\"container_2\">\n",
      "\n",
      "                </ul>\n",
      "\n",
      "              </div>\n",
      "\t      <FORM ENCTYPE=\"MULTIPART/FORM-DATA\" METHOD=\"POST\" ACTION=\"/sift-bin/SIFT_seq_submit2.pl\" id=\"Form\">\n",
      "              <div id=\"middle_content_template\">\n",
      "\t        <div id=\"tool_instructions\">\n",
      "\t\t  <div id=\"instruct_title\">SIFT Sequence</div>\n",
      "\n",
      "                  <p><b>SIFT Sequence provides SIFT predictions for a given protein FASTA sequence.</b> This will take 10-15 min because we must search your protein sequence against a database to pick the related sequences. You can also <A HREF=\"SIFT_related_seqs_submit.html\">submit your protein sequence and related sequences</A> or <A HREF=\"SIFT_aligned_seqs_submit.html\">aligned sequences </A> if you already have them.<BR>\n",
      " \n",
      "\t\t\t<BR>\n",
      "\t\t\t<i>Results are deleted after 24 hours, so please save them! </i>\n",
      "\t\t\t<BR>\n",
      "\t\t\t[<A HREF=\"SIFT_help.html#CHANGING_PARAMETERS\">Preventing connection failures</A>]\n",
      "\n",
      "\n",
      "\t\t  </p>\n",
      "\n",
      "\t\t</div>\n",
      "\n",
      "\t\t\t\n",
      "\t\t\t<HR>\n",
      "\t\t\t<font color=blue>Protein sequence</font><BR>\n",
      "\t\t\t<BR>\n",
      "\t\t\tName of file containing protein query sequence\n",
      "\t\t\t<A HREF=\"http://www.ncbi.nlm.nih.gov/BLAST/fasta.shtml\">(fasta format)</A>.<BR>\n",
      "\t\t\t<INPUT SIZE=50 NAME=\"query_file\" TYPE=\"file\"><BR><BR>\n",
      "\t\t\t<center><font size=\"6\" color=\"blue\">-or-</font></center><BR>\n",
      "\t\t\tPaste in your protein query sequence \n",
      "\t\t\t(<A HREF=\"javascript:ChgText()\" >Upload example</A>)\n",
      "\t\t\t<A HREF=\"http://www.ncbi.nlm.nih.gov/BLAST/fasta.shtml\">(fasta format)</A>.<BR>\n",
      "\t\t\t<TEXTAREA NAME = \"queryseq\" id=\"queryseq\" ROWS=10 COLS=70></TEXTAREA>\n",
      "\t\t\t<BR>\n",
      "\t\t\t<BR>\n",
      "\t\t\t<HR>\n",
      "\t\t\t<BR>\n",
      "\t\t\tEnter the substitutions of interest <A HREF=\"SIFT_help.html#substitution\">[format]</A>:<BR>\n",
      "\t\t\t<TEXTAREA NAME = \"substitutions\" id=\"substitutions\" ROWS=5 COLS=50></TEXTAREA>\n",
      "\t\t\t<BR><font size=\"6\" color=\"blue\">-or-</font>\n",
      "\n",
      "\n",
      "\t\t\t<BR>\n",
      "\t\t\tUpload a file containing substitutions of interest <A HREF=\"SIFT_help.html#substitution\">[format]</A>:<br>\n",
      "                        <INPUT SIZE=50 NAME=\"substitution_file\" TYPE=\"file\"><BR><BR>\n",
      "\n",
      "\t\t\t<HR>\n",
      "\t\t\t<BR>\n",
      "\t\t\t<font color=blue>Parameters</font><BR>\n",
      "\t\t\t<BR>\n",
      "\t \t\tSelect database to search:\n",
      "  \t\t\t<SELECT NAME=\"database\" SIZE=\"1\">\n",
      "  \t\t\t<OPTION VALUE=\"sp\" SELECTED>UniProt-SwissProt 2010_09 \n",
      "  \t\t\t<OPTION VALUE=\"sp_tr\">UniProt-SwissProt + TrEMBL 2010_09 \n",
      "  \t\t\t<OPTION VALUE=\"ur_ninety\">UniRef90 2011_08\n",
      "<!-- \t\t\t<OPTION VALUE=\"nr\">NCBI nonredundant 2010 Sept --> \n",
      "  \t\t\t</SELECT>\n",
      "\t\t\t<BR><A HREF=\"SIFT_help.html#CHOOSING_DATABASE\">Warnings for you predictions? Change this option</A><BR>\n",
      "\t\t\t<BR>\n",
      "\t\t\t<BR>\n",
      "\t\t\tMedian conservation of sequences <INPUT TYPE=\"TEXT\" SIZE=5 NAME=\"info\" VALUE=3.00> <BR>\n",
      "\t\t\t<i>(between 0 and 4.32, we recommend between 2.75 and 3.25)</i><BR>\n",
      "\t\t\t<BR>\n",
      "\t\t\tRemove sequences more than\n",
      "\t\t\t<input type=text name=seq_identity_filter size=3 maxlength = 3 value=\"90\">\n",
      "\t\t\tpercent identical to query<BR>\n",
      "\t\t\t<i>This cleans out any polymorphic alleles (sequences in the database that are nearly identical to your protein but already containing the substitutions of interest). The contaminating sequences may cause the substitution to be predicted as tolerated.</i>\n",
      " <p>\n",
      "                        Enter your email address if you want the results through email : (Optional)<BR>\n",
      "\n",
      "                        <INPUT SIZE=50 NAME=\"address\">\n",
      "                        </INPUT SIZE>\n",
      "                        <BR>\n",
      "<HR>\n",
      "\t\t\t<BR>\n",
      "\t\t\t<!--<INPUT TYPE=submit>-->\n",
      "                <input type=\"button\" id=\"submit_button\" value=\"Submit\"></input>\n",
      "\n",
      "                <INPUT TYPE=reset>\n",
      "\n",
      "\n",
      "\t\t  </p>\n",
      "\t\t  </FORM>\n",
      "              </div>\n",
      "            </div>\n",
      "          </div>\n",
      "\n",
      "        </div>\n",
      "      </div>\n",
      "    </div>\n",
      "    <div id=\"home_border\">\n",
      "      <div id=\"home_bar\">\n",
      "        <div id=\"home_container\">\n",
      "        </div>\n",
      "      </div>\n",
      "\n",
      "      <div id=\"footer_spacer\"></div>\n",
      "    </div>\n",
      "\n",
      "<P>\n",
      "  <script src=\"/javascripts/footer.js\" type=\"text/javascript\"> </script>\n",
      "\n",
      "  </body>\n",
      "</html>\n",
      "\n",
      "<script>\n",
      "$( document ).ready(function() {\n",
      "       $(\"div.modal-overlay\").hide();\n",
      "      $(\".loading\").hide();\n",
      "});\n",
      "\n",
      "$('#submit_button').on('click', function(e) {\n",
      "    \n",
      "  \n",
      "     var tag_inputs = document.getElementsByTagName('input');\n",
      "     var textarea = document.getElementsByTagName ('textarea');\n",
      "     //inputs = inputs.concat (textarea);\n",
      "     var inputs = [];\n",
      "     for (var index = 0; index < tag_inputs.length; index++) {\n",
      "        inputs.push (tag_inputs[index]);\n",
      "     }\n",
      "     for (var index = 0; index < textarea.length; index++) {\n",
      "        inputs.push (textarea[index]);\n",
      "     }\n",
      "\n",
      "\n",
      "     for(var x = 0; x < inputs.length; x++) {\n",
      "           \n",
      "       // alert (\"checking \" + inputs[x].name);\n",
      "             if (inputs[x].name == \"query_file\") {\n",
      "                 if (!checkFileOK (inputs[x].value, ['.fasta', '.fa'], false)) {\n",
      "                    alert (\"File must end in .fasta or .fa\");\n",
      "                     return false;\n",
      "                 }\n",
      "              //   if (inputs[x].files.length > 0 && inputs[x].files[0].size > 5000000) {\n",
      "                  if (!checkFileSize (inputs[x].files)) {\n",
      "                      alert (\"Files should be < 5 MB.\");\n",
      "                      return false;\n",
      "                 }\n",
      "\n",
      "             } else if (inputs[x].name == \"queryseq\") {\n",
      "                // alert (\"check QuerySeq\" + inputs[x].value);\n",
      "                 if (!validateFasta (inputs[x].value)) {\n",
      "                     alert (\"Please put your sequence in fasta format\");\n",
      "                     return false;\n",
      "                 }\n",
      "             } else if (inputs[x].name == \"substitutions\") {\n",
      " //                alert (\"check substitutions\");\n",
      "                 if (!validateSubstitutions (inputs[x].value)) {\n",
      "                     alert (\"The substitutions are not in correct format\");\n",
      "                     return false;\n",
      "                 }\n",
      "             } else if (inputs[x].name == \"substitution_file\") {\n",
      "                    if (!checkFileOK (inputs[x].value, ['.subst', '.txt'], false)) {\n",
      "                        alert (\"File must end in .subst or .txt.  File size cannot exceed 5 MB.\");\n",
      "                     return false;\n",
      "                    }\n",
      "                    if ( !checkFileSize (inputs[x].files) ) {\n",
      "                        alert (\"Files should be < 5 MB.\");\n",
      "                        return false;\n",
      "                    }\n",
      "             } else if (inputs[x].name == \"info\") {\n",
      "                 if (!validateMedianSeqInfo (inputs[x].value, true)) {\n",
      "                     alert (\"Please set the median conservation of sequences between 0 and 4.32  . How about 3.00?\");\n",
      "                     return false;\n",
      "                 }\n",
      "                 //alert (\"check substitutions\"); // between 0 and 4.32 (this is a dable)\n",
      "             } else if (inputs[x].name == \"seq_identity_filter\") {\n",
      "                // alert (\"check substitutions\"); // string between 0 and 1[[\n",
      "                if (!validateSeqIdentity (inputs[x].value, true)) {\n",
      "                    alert (\"Sequence identity must be between 0 and 100. How about 90?\")\n",
      "                    return false;\n",
      "                }\n",
      "             } else if (inputs[x].name == \"address\") {\n",
      "                if (!validateEmail (inputs[x].value)) {\n",
      "                     alert (\"Please correct your email address\");\n",
      "                     return false;\n",
      "                 }\n",
      "             }\n",
      "          \n",
      "     } // end for loop\n",
      "     //alert (document.getElementById (inputs[x].name).validationMessage);\n",
      "\n",
      "//     $(\"div.modal-overlay\").show();\n",
      " //    $(\".loading\").show();\n",
      "\n",
      "\n",
      "     $(\"#Form\").submit();\n",
      "\n",
      "    \n",
      "\n",
      "       e.preventDefault();\n",
      " });\n",
      "</script>\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "\n",
    "def run_sift_analysis(protein_sequence):\n",
    "    url = \"https://sift.bii.a-star.edu.sg/www/SIFT_seq_submit2.html\"\n",
    "    params = {\n",
    "        \"seq\": protein_sequence,\n",
    "        \"siftSubmit\": \"Submit\",\n",
    "        \"siftseqtxt\": \"\",\n",
    "        \"siftupload\": \"\",\n",
    "    }\n",
    "    response = requests.post(url, data=params)\n",
    "    # Process the response and extract relevant information\n",
    "    # (e.g., SIFT scores or predicted impact of variants)\n",
    "    # In this example, we'll simply print the response text\n",
    "    print(response.text)\n",
    "\n",
    "# Define the normal and SNP-containing protein sequences\n",
    "normal_protein_sequence = \"MVHLTPVEKSAVTALWGKVNVDEVGGEALGRLLVVYPWTQRFFASFE\"\n",
    "snp_protein_sequence = \"MVHLTPVEKSAVTALWGKVNVDEVGGEALGRLLVVYPWTQRFFASFE\"  # Example SNP\n",
    "\n",
    "# Run SIFT analysis for both sequences\n",
    "print(\"Normal protein sequence:\")\n",
    "run_sift_analysis(normal_protein_sequence)\n",
    "\n",
    "print(\"\\nSNP-containing protein sequence:\")\n",
    "run_sift_analysis(snp_protein_sequence)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To use the SIFT tool to predict the impact of a SNP (Single Nucleotide Polymorphism) within the PDC gene on the PDC protein function, you would typically follow these steps:\n",
    "\n",
    "Identify the SNP: Choose a specific SNP within the PDC gene sequence.\n",
    "Retrieve Protein Sequence: Obtain the normal and SNP-containing amino acid sequences of the PDC protein.\n",
    "Submit Sequences to SIFT: Use the SIFT web server to submit the normal and SNP-containing protein sequences for analysis.\n",
    "Analyze SIFT Results: Interpret the SIFT results to determine the impact of the SNP on the protein function.\n",
    "Below is an example Python code demonstrating these steps:\n",
    "\n",
    "python\n",
    "Copy code\n",
    "import requests\n",
    "\n",
    "def run_sift_analysis(protein_sequence):\n",
    "    url = \"https://sift.bii.a-star.edu.sg/www/SIFT_seq_submit2.html\"\n",
    "    params = {\n",
    "        \"seq\": protein_sequence,\n",
    "        \"siftSubmit\": \"Submit\",\n",
    "        \"siftseqtxt\": \"\",\n",
    "        \"siftupload\": \"\",\n",
    "    }\n",
    "    response = requests.post(url, data=params)\n",
    "    # Process the response and extract relevant information\n",
    "    # (e.g., SIFT scores or predicted impact of variants)\n",
    "    # In this example, we'll simply print the response text\n",
    "    print(response.text)\n",
    "\n",
    "# Define the normal and SNP-containing protein sequences\n",
    "normal_protein_sequence = \"MVHLTPVEKSAVTALWGKVNVDEVGGEALGRLLVVYPWTQRFFASFE\"\n",
    "snp_protein_sequence = \"MVHLTPVEKSAVTALWGKVNVDEVGGEALGRLLVVYPWTQRFFASFE\"  # Example SNP\n",
    "\n",
    "# Run SIFT analysis for both sequences\n",
    "print(\"Normal protein sequence:\")\n",
    "run_sift_analysis(normal_protein_sequence)\n",
    "\n",
    "print(\"\\nSNP-containing protein sequence:\")\n",
    "run_sift_analysis(snp_protein_sequence)\n",
    "In this code:\n",
    "\n",
    "run_sift_analysis function sends a protein sequence to the SIFT web server for analysis.\n",
    "We define the normal protein sequence and a hypothetical SNP-containing protein sequence (you need to replace this with the actual sequences).\n",
    "We then run the SIFT analysis for both sequences and print the response text, which contains the SIFT scores or predicted impact of variants.\n",
    "Remember to replace the example sequences with the actual sequences obtained from the PDC gene and the specific SNP you want to analyze. Additionally, you may need to parse the response from the SIFT server to extract and interpret the relevant information, such as SIFT scores or predictions.\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
