{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genome sequence downloaded and saved to yeast_genome.fasta\n",
      "Accession number: NG_013229.2\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez, SeqIO\n",
    "\n",
    "def download_and_save_chromosomes(accession_numbers, output_file):\n",
    "    # Set your email for Entrez\n",
    "    Entrez.email = \"your_email@example.com\"\n",
    "\n",
    "    # List to store the downloaded sequences\n",
    "    chromosome_sequences = [] \n",
    "    # Loop through each accession number and download the sequence\n",
    "    for accession_number in accession_numbers:\n",
    "        handle = Entrez.efetch(db=\"nucleotide\", id=accession_number, rettype=\"fasta\", retmode=\"text\")\n",
    "        record = SeqIO.read(handle, \"fasta\")\n",
    "        handle.close()\n",
    "\n",
    "        # Append the sequence to the list\n",
    "        chromosome_sequences.append(record)\n",
    "\n",
    "    # Save all sequences to a single file\n",
    "    with open(output_file, \"w\") as f:\n",
    "        SeqIO.write(chromosome_sequences, f, \"fasta\")\n",
    "\n",
    "    print(f\"Chromosomes downloaded and saved to {output_file}\")\n",
    "    return chromosome_sequences\n",
    "\n",
    "accession_list = [\"NC_001133.9\", \"NC_001134.8\", \"NC_001135.5\", \"NC_001136.10\", \"NC_001137.3\",\n",
    "                 \"NC_001138.5\", \"NC_001139.9\", \"NC_001140.6\", \"NC_001141.2\", \"NC_001142.9\",\n",
    "                 \"NC_001143.9\", \"NC_001144.5\", \"NC_001145.3\", \"NC_001146.8\", \"NC_001147.5\", \"NC_001148.4\"]\n",
    "output_file = \"yeast_genome.fasta\"\n",
    "\n",
    "# Download and save chromosomes to a single file\n",
    "download_and_save_chromosomes(accession_list, output_file)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ARS sequence ((A/T)TTTA(C/T)(A/G)TTT(A/T)) found at positions (1-based indexing): [159, 423, 10087, 23067, 24105, 36462, 38702, 55048, 82587, 90507, 101874, 108868]\n",
      "Matching entries:\n",
      "Sequence: TTTTATATTTTTAACTTTTCCCACAAT - Start Index: 159\n",
      "Sequence: ATTTATATTTTATTCAGGTCCAGGATT - Start Index: 423\n",
      "Sequence: TTTTATATTTTTAGTAGAGACATTGTT - Start Index: 10087\n",
      "Sequence: TTTTATGTTTTTGTGGAGGCAAGGTCT - Start Index: 23067\n",
      "Sequence: TTTTATATTTTTTCAGAGAAAGGGTCT - Start Index: 24105\n",
      "Sequence: ATTTATATTTAATGATATTCAGTATGA - Start Index: 36462\n",
      "Sequence: TTTTATATTTTTAGTAGAGACGGGGTT - Start Index: 38702\n",
      "Sequence: TTTTACATTTTGATGGCTCCTATACAC - Start Index: 55048\n",
      "Sequence: ATTTATGTTTTTATAAGCTGGTACATT - Start Index: 82587\n",
      "Sequence: ATTTACATTTTTAAAATATTGTGTGTT - Start Index: 90507\n",
      "Sequence: TTTTACATTTTTTAATGGTTAAAAAAA - Start Index: 101874\n",
      "Sequence: TTTTATGTTTTTTAAAAATGTATGTAG - Start Index: 108868\n",
      "First Origin of Replication (ORI) position: 159\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "from Bio import SeqIO\n",
    "from Bio import Entrez\n",
    "\n",
    "def find_ori_using_ars_for_all_chromosomes(file_path):\n",
    "    # Read the FASTA file containing all chromosomes\n",
    "    records = SeqIO.to_dict(SeqIO.parse(file_path, \"fasta\"))\n",
    "\n",
    "    # Initialize a dictionary to store results for each chromosome\n",
    "    ori_results = {}\n",
    "\n",
    "    for chromosome_id, record in records.items():\n",
    "        # Get the DNA sequence for the current chromosome\n",
    "        dna_seq = str(record.seq)\n",
    "\n",
    "        # Define ARS sequence pattern\n",
    "        ars_seq = \"(A/T)TTTA(C/T)(A/G)TTT(A/T)\"\n",
    "        ars_pattern = re.compile(\"[AT]TTTA[CT][AG]TTT[AT]\")\n",
    "\n",
    "        # Find ARS sequence matches\n",
    "        ars_matches = [match.start() for match in ars_pattern.finditer(dna_seq)]\n",
    "\n",
    "        matching_entries = []\n",
    "        first_ori_position = None\n",
    "\n",
    "        # Process matching entries for the current chromosome\n",
    "        if ars_matches:\n",
    "            print(f\"Chromosome {chromosome_id}: ARS sequence ({ars_seq}) found at positions (1-based indexing): {ars_matches}\")\n",
    "\n",
    "            for ori_posn in ars_matches:\n",
    "                matching_entry = record[ori_posn: ori_posn + len(ars_seq)]\n",
    "                matching_entries.append((matching_entry.id, str(matching_entry.seq), ori_posn))\n",
    "\n",
    "                if first_ori_position is None:\n",
    "                    first_ori_position = ori_posn\n",
    "\n",
    "        else:\n",
    "            print(f\"Chromosome {chromosome_id}: No ARS sequence found.\")\n",
    "\n",
    "        # Store results for the current chromosome in the dictionary\n",
    "        ori_results[chromosome_id] = {\n",
    "            'matching_entries': matching_entries,\n",
    "            'first_ori_position': first_ori_position\n",
    "        }\n",
    "\n",
    "    return ori_results\n",
    "\n",
    "# Usage\n",
    "file_path = \"yeast_genome.fasta\"\n",
    "ori_results = find_ori_using_ars_for_all_chromosomes(file_path)\n",
    "\n",
    "# Print results for each chromosome\n",
    "for chromosome_id, results in ori_results.items():\n",
    "    matching_entries = results['matching_entries']\n",
    "    first_ori_position = results['first_ori_position']\n",
    "\n",
    "    if matching_entries:\n",
    "        print(f\"\\nChromosome {chromosome_id}: Matching entries:\")\n",
    "        for entry_id, sequence, ori_posn in matching_entries:\n",
    "            print(f\"  Sequence: {sequence} - Start Index: {ori_posn}\")\n",
    "\n",
    "        print(f\"  First Origin of Replication (ORI) position: {first_ori_position}\")\n",
    "    else:\n",
    "        print(f\"\\nChromosome {chromosome_id}: No matching entries found.\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "- Reference:\n",
    "1. https://en.wikipedia.org/wiki/Autonomously_replicating_sequence-:~:text=An%20autonomously%20replicating%20sequence%20(ARS,their%20effect%20on%20plasmid%20stability.\n",
    "2. https://pubmed.ncbi.nlm.nih.gov/9441849/#:~:text=Autonomously%20replicating%20sequence%20(ARS)%20elements%20were%20first%20identified%20in%20the,extrachromosomal%20maintenance%20of%20plasmid%20DNA.\n",
    "3. https://www.sciencedirect.com/topics/medicine-and-dentistry/autonomously-replicating-sequence\n",
    "4. https://en.wikipedia.org/wiki/Baker%27s_yeast\n",
    "5. https://en.wikipedia.org/wiki/Autonomously_replicating_sequence\n",
    "6. https://www.ncbi.nlm.nih.gov/search/all/?term=Saccharomyces+cerevisiae\n",
    "- using the hint given in the q, (S. cerevisiae has an Autonomously Replicating Sequence with features you can search.)\n",
    "- the ARS that contains the origin of replication in the yeast genome is - \"(A/T)TTTA(C/T)(A/G)TTT(A/T)\" found in Saccharomyces cerevisiae which i derived from the sequence given from the above links\n",
    " where \n",
    "- the \"/\" slash shows that it could be either the first base or the second one which will be checked by the code for matching the seq from yeastgenome.gb file.\n",
    "\n",
    "- \"T\" is thymine.\n",
    "- \"A\" is adenine\n",
    "- \"C\" is cytosine (C) \n",
    "- \"G\" is guanine (G)\n",
    "\n",
    "- So after runnin above code ouput is coming out to be-\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "- This means that there are three occurences matching the ars_seq \"(A/T)TTTA(C/T)(A/G)TTT(A/T)\" and this seq is present in three different chromosomes of the yeast genome, starting at positions 56946, 65503, and 68265.\n",
    "- The first copy starts at 56946 which after crosschecking from yeast_genome.gb file, is correct- TTTTATGTTTA which matches our ars sequence.\n",
    " \n",
    " this is a short snippet frm yeast_genome.gb file-\n",
    "\n",
    "     56941 ctgtat'tttt atgttta'att ataacccctt taggattata atttaaatta atttaaatat\n",
    " \n",
    " where 56941 denoted the first nucleotides seq starting idx(1 based) and now if we use slicing just to crosscheck our ans \n",
    " \"ttttatgttta\" is extracted , which starts on the \"t\" at position 56946.\n",
    "\n",
    "- Then i used the nt_search function as done in tuts - from Bio.SeqUtils module to search for occurrences of ars_seq \"(A/T)TTTA(C/T)(A/G)TTT(A/T)\" in the DNA sequence.\n",
    "\n",
    "- If matches are found, i store them in \"matching_entries\" list and also their coresponding position from the .gb file\n",
    "\n",
    "- finally for the ans, I hv taken the first position as the Origin of Replication (ori).\n",
    " "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
